plutus-playground/plutus-playground-server/src/Playground/Interpreter.hs

```diff
@@ -1,166 +1,195 @@
-    loadSource fileName $ \moduleName -> do
-        exports <- getModuleExports moduleName
-        walletFunctions <- catMaybes <$> traverse isWalletFunction exports
-        schemas <- traverse getSchema walletFunctions
-        pure (schemas <> pure payToPublicKey_Schema)
-
-jsonToString :: ToJSON a => a -> String
-jsonToString = show . JSON.encode
-
-{-# ANN getSchema ("HLint: ignore" :: String) #-}
-
-getSchema ::
-       (MonadInterpreter m, MonadError PlaygroundError m)
-    => ModuleElem
-    -> m (FunctionSchema Schema)
-getSchema (Fun m) = interpret m (as :: FunctionSchema Schema)
-getSchema _       = throwError FunctionSchemaError
-  -- error "Trying to get a schema by calling something other than a function"
-
-{-# ANN getJsonString ("HLint: ignore" :: String) #-}
-
-getJsonString :: (MonadError PlaygroundError m) => JSON.Value -> m String
-getJsonString (JSON.String s) = pure $ Text.unpack s
-getJsonString v =
-    throwError . DecodeJsonTypeError "String" . BSL.unpack . encode $ v
+compile source = do
+    avoidUnsafe source
+    withSystemTempFile "Main.hs" $ \file handle -> do
+        (exitCode, stdout, stderr) <-
+            runscript handle file $ mkCompileScript (Newtype.unpack source)
+        result <-
+            case exitCode of
+                ExitSuccess -> pure stdout
+                _ ->
+                    throwError . CompilationErrors $
+                    parseErrorsText (Text.pack stderr)
+        let eSchema = JSON.eitherDecodeStrict . BS8.pack $ result
+        case eSchema of
+            Left err ->
+                throwError . OtherError $
+                "unable to decode compilation result" <> err
+            Right schema -> pure schema
 
 runFunction ::
-       (MonadInterpreter m, MonadError PlaygroundError m)
+       (MonadMask m, MonadIO m, MonadError PlaygroundError m)
     => Evaluation
     -> m (Blockchain, [EmulatorEvent], [(Wallet, Value)])
 runFunction evaluation = do
-    avoidUnsafe $ sourceCode evaluation
+    let source = sourceCode evaluation
+    avoidUnsafe source
     expr <- mkExpr evaluation
-    fileName <- writeTempSource $ sourceCode evaluation
-    loadSource fileName $ \_ -> do
-        setImportsQ
-            [ ("Playground.Interpreter.Util", Nothing)
-            , ("Playground.API", Nothing)
-            , ("Wallet.Emulator", Nothing)
-            , ("Ledger.Types", Nothing)
-            , ("Data.Map", Nothing)
-            ]
-        liftIO . putStrLn $ expr
-        res <-
-            interpret
-                expr
-                (as :: Either PlaygroundError ( Blockchain
-                                              , [EmulatorEvent]
-                                              , [(Wallet, Value)]))
-        case res of
-            Left e  -> throwM . UnknownError $ show e
-            Right r -> pure r
-
-mkExpr :: (MonadError PlaygroundError m) => Evaluation -> m String
+    withSystemTempFile "Main.hs" $ \file handle -> do
```

Comment:
fair bit of duplication with `compile`

Reply:
I removed as much as hlint told me to. I felt there were too many parameters to pull it out into another function, I don't think it would be as clear what is going on.